File: app/src/main/java/com/radardeal/app/core/Formatters.kt
package com.radardeal.app.core

import java.util.Locale
import kotlin.math.abs
import kotlin.math.roundToInt

/**
 * Small, allocation-cheap formatting helpers shared by the UI and the notification layer.
 *
 * They are deliberately null-tolerant: a Vinted listing can be missing its price, its size or
 * its brand, and no screen may ever crash or show "null" because of it.
 */
object Formatters {

    /** "42 €", "39,50 €", or the [fallback] when the price is unknown. */
    fun price(amount: Double?, currency: String? = "EUR", fallback: String = "Prix inconnu"): String {
        if (amount == null || amount.isNaN() || amount.isInfinite()) return fallback
        val symbol = currencySymbol(currency)
        val rounded = (amount * 100).roundToInt() / 100.0
        val text = if (abs(rounded - rounded.roundToInt()) < 0.005) {
            rounded.roundToInt().toString()
        } else {
            String.format(Locale.FRANCE, "%.2f", rounded)
        }
        return "$text $symbol"
    }

    fun currencySymbol(currency: String?): String = when (currency?.uppercase(Locale.ROOT)) {
        null, "", "EUR" -> "€"
        "GBP" -> "£"
        "USD" -> "$"
        "PLN" -> "zł"
        "CZK" -> "Kč"
        "SEK" -> "kr"
        "HUF" -> "Ft"
        else -> currency
    }

    /** "Il y a 18 sec", "Il y a 4 min", "Il y a 3 h", "Il y a 2 j". */
    fun relativeTime(timestampMs: Long?, nowMs: Long = System.currentTimeMillis()): String {
        if (timestampMs == null || timestampMs <= 0L) return "Jamais"
        val delta = nowMs - timestampMs
        if (delta < 0) return "À l'instant"
        val seconds = delta / 1000
        return when {
            seconds < 10 -> "À l'instant"
            seconds < 60 -> "Il y a $seconds sec"
            seconds < 3600 -> "Il y a ${seconds / 60} min"
            seconds < 86_400 -> "Il y a ${seconds / 3600} h"
            seconds < 2_592_000 -> "Il y a ${seconds / 86_400} j"
            else -> "Il y a ${seconds / 2_592_000} mois"
        }
    }

    /** Compact countdown used on the watch list: "30 sec", "2 min", "1 h". */
    fun duration(seconds: Int): String = when {
        seconds < 60 -> "$seconds sec"
        seconds % 3600 == 0 -> "${seconds / 3600} h"
        seconds < 3600 -> "${seconds / 60} min"
        else -> "${seconds / 60} min"
    }

    /** "-32 %" for a 0.32 ratio. Always returns a signed, rounded integer percentage. */
    fun percent(ratio: Double?): String {
        if (ratio == null || ratio.isNaN() || ratio.isInfinite()) return "—"
        val pct = (ratio * 100).roundToInt()
        return "$pct %"
    }
}

File: ios/RadarDealCore/Sources/RadarDealCore/Formatters.swift
import Foundation

/// Small formatting helpers shared by every screen.
///
/// They are deliberately optional-tolerant: a Vinted listing can be missing its price, its size
/// or its brand, and no screen may ever crash or show "nil" because of it.
public enum Formatters {

    /// "42 €", "39,50 €", or `fallback` when the price is unknown.
    public static func price(
        _ amount: Double?,
        currency: String? = "EUR",
        fallback: String = "Prix inconnu"
    ) -> String {
        guard let amount, amount.isFinite else { return fallback }

        let symbol = currencySymbol(currency)
        let rounded = (amount * 100).rounded() / 100

        if abs(rounded - rounded.rounded()) < 0.005 {
            return "\(Int(rounded.rounded())) \(symbol)"
        }
        // French convention: comma as the decimal separator, independent of device locale so
        // the app reads the same everywhere.
        let text = String(format: "%.2f", rounded).replacingOccurrences(of: ".", with: ",")
        return "\(text) \(symbol)"
    }

    public static func currencySymbol(_ currency: String?) -> String {
        switch currency?.uppercased() {
        case nil, "", "EUR": return "€"
        case "GBP": return "£"
        case "USD": return "$"
        case "PLN": return "zł"
        case "CZK": return "Kč"
        case "SEK": return "kr"
        case "HUF": return "Ft"
        case let other?: return other
        }
    }

    /// "Il y a 18 sec", "Il y a 4 min", "Il y a 3 h", "Il y a 2 j".
    public static func relativeTime(_ date: Date?, now: Date = Date()) -> String {
        guard let date, date.timeIntervalSince1970 > 0 else { return "Jamais" }

        let delta = now.timeIntervalSince(date)
        if delta < 0 { return "À l'instant" }

        let seconds = Int(delta)
        switch seconds {
        case ..<10: return "À l'instant"
        case ..<60: return "Il y a \(seconds) sec"
        case ..<3600: return "Il y a \(seconds / 60) min"
        case ..<86_400: return "Il y a \(seconds / 3600) h"
        case ..<2_592_000: return "Il y a \(seconds / 86_400) j"
        default: return "Il y a \(seconds / 2_592_000) mois"
        }
    }

    /// Compact interval used on the watch list: "30 sec", "2 min", "1 h".
    public static func duration(seconds: Int) -> String {
        if seconds < 60 { return "\(seconds) sec" }
        if seconds % 3600 == 0 { return "\(seconds / 3600) h" }
        return "\(seconds / 60) min"
    }

    /// "32 %" for a 0.324 ratio.
    public static func percent(_ ratio: Double?) -> String {
        guard let ratio, ratio.isFinite else { return "—" }
        return "\(Int((ratio * 100).rounded())) %"
    }
}
